### PR TITLE
[Color system] Auto generate inverse, light, and dark variants

### PR DESCRIPTION
**Pull request by tmlayton**
### WHY are these changes introduced?

TL;DR To simplify the color adjustments file, deduplicate, and automate otherwise manual work.

In some cases the inverse, light, or dark variant of a color is needed. The two examples below are 1) some button text should always be white regardless of light or dark mode and 2) toast uses inverse colors.

Instead of manually specifying the role name and adjustments for light, dark, and inverse variants, they can be generated from the base role since the dark and light adjustments are already known.

### WHAT is this pull request doing?

Too much.

- Auto generates dark, light, and inverse role variants from the base role and removes the manual entries from the color adjustments.
- Adds docs to code
- Changes schema of variants file
- Updates naming
- Changes role `branded` to `primary`
- Adds script to generate color docs (`yarn colordocs`)
- Updates color docs

### Trade-offs

- This will currently generate variants for every role, so it is potentially a lot of extra colors which are not needed. Although the usage of these variants is still foggy.
- Some control is lost over the role name as it simply adds a suffix to the base role name, for example `textOnDark` becomes `textOnSurfaceDark` as the base role is `textOnSurface`
- A bit more magic which can be confusing

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}

```

</details>

### 🎩 checklist

* [ ] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [ ] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
* [ ] Updated the component's `README.md` with documentation changes
* [ ] [Tophatted documentation](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting%20documentation.md) changes in the style guide
* [ ] For visual design changes, pinged one of @ HYPD, @ mirualves, @ sarahill, or @ ry5n to update the [Polaris UI kit](https://polaris.shopify.com/resources/polaris-ui-kit)

### CODE REVIEW

**Comment by tmlayton** [comment]
Need to update docs, too.

**Comment by martenbjork** [comment]
This solution is more elegant, but it's getting increasingly hard to understand how the color factory works. The number of inputs and and outputs are starting to feel a bit unpredictable. Having the super verbose adjustments.ts has its benefits:

- It's just easy to grasp. (n) adjustments -> factory -> (n) outputs
- You can extend each color separately. If we want to add, say, `description` or `tags` to colors, we can easily do it for each color. (Having this for each color could be useful for Figma plugins, documentation etc.)
- If you're building tooling, you can easily iterate over the adjustments file and get the names and values of all colors and enable tweaking. After this PR, you would have to iterate over the colors in the adjustments file, then run the factory and then figure out which ones were generated automatically.

**Comment by dleroux** [comment]
I see where Marten is coming from. I wonder if this is a bit of premature optimization. While we are developing and sorting out the final it might be best that we do have a more explicit list of values? 

I do however like this tradeoff as it makes it clearer:

- Some control is lost over the role name as it simply adds a suffix to the base role name, for example textOnDark becomes textOnSurfaceDark as the base role is textOnSurface.

File: src/components/Frame/components/Toast/Toast.scss [hunk 1]
```diff
@@ -10,11 +10,11 @@ $Backdrop-opacity: 0.88;
   padding: spacing(tight) spacing();
   border-radius: var(--p-border-radius-wide, border-radius());
   background: var(
-    --p-surface-inverse-foreground,
+    --p-surface-foreground-inverse,
```
Comment (danrosenthal) [inline src/components/Frame/components/Toast/Toast.scss]: I'm okay with the name changes here. It actually feels clearer to me what they're for with this change.

**Comment by tmlayton** [comment]
> Would it be possible to rethink this a bit, adding some sort of configuration flag in each object in `color-adjustments` that would indicate we should generate these for that variation? Or perhaps we could somehow abstract that up to the color role level.

We chatted about this directly, just wanted to share with everyone here. I agree with not wanting to create unneeded CSS variables, but in theory all are needed. This allows any consumer to build a fixed dark, light, or inverse dialog with any component and the extra CSS variables really don’t have a negative impact.

**Comment by tmlayton** [comment]
Needs a rebase, otherwise it is ready for review. The last commit changed the schema for the role variants and updated naming and types.

**Comment by tmlayton** [comment]
> Having the super verbose adjustments.ts has its benefits

Totally agree, and I think it is still verbose with the added benefit of no longer being redundant. 

> It's just easy to grasp. (n) adjustments -> factory -> (n) outputs

I would argue this makes the adjustments file even easier to grasp. It boils that file down to its essence, nothing superfluous which means all signal and no noise.

> You can extend each color separately. If we want to add, say, description or tags to colors, we can easily do it for each color. (Having this for each color could be useful for Figma plugins, documentation etc.)

This is still possible. The ability to extend the inverse, dark, and light variants independently is lost, but they are merely children of their parent variant (or in other words variants of a variant). I would argue dark, light, and inverse variants should _never_ be extended independently, rather should be thought of as a concept which universally applies to all variants which have a color that is different in light and dark mode.

> If you're building tooling, you can easily iterate over the adjustments file and get the names and values of all colors and enable tweaking. After this PR, you would have to iterate over the colors in the adjustments file, then run the factory and then figure out which ones were generated automatically.

I think this ties back to the previous point. Dark, light, and inverse variants are not independent concepts. They are derived from the parent variant’s light and dark HSLA settings. Iterating over the ~adjustments~  variants file still produces every color created from the factory and each unique variant.

> premature optimization

Premature optimization has to do with making software faster that isn’t yet slow in an effort to reduce wasted effort, so I don’t think it applies here. This is refactoring and deduplication.

File: src/utilities/theme/types.ts [hunk 2]
```diff
@@ -11,6 +11,38 @@ export interface ThemeLogo {
...
+  surface: string;
+  /** The onSurface role is made up of elements which appear on top of a surface, including borders, neutral icons, and text. When a light surface is provided, onSurface values will be dark. When a dark surface is provided, onSurface values will be light. */
+  onSurface: string;
+  /** The interactive role is used to express interactivity in components. It is used in links, as an indicator of focus, and as an indicator of selected interactive states. */
+  interactive: string;
+  /** A neutral interactive color role, for use in secondary and tertiary buttons as a background color, as well as in form elements as a background color. */
+  neutral: string;
+  /** A branded interactive color, for use in primary buttons as a background color. Also used in navigation and tabs for icons, and for a surface color when in a selected state. */
+  branded: string;
+  /** Used to communicate destructive outcomes on interactive elements, for communicating errors, and to indicate a critical event in inert elements that requires immediate merchant action. */
```
[the remark this answers is not in the document]
Comment (tmlayton) [reply src/utilities/theme/types.ts]: Good question, for now this is still product-focused and the docs are very much WIP.

**Comment by tmlayton** [comment]
> the nested reducers are getting a little out of hand

Yes, not great as the readability is low. The factory will be refactored again to support child theme providers and would be a good time to see if some of this can be flattened.

File: src/utilities/theme/types.ts [hunk 3]
```diff
@@ -11,6 +11,38 @@ export interface ThemeLogo {
   width?: number;
 }
 
+export type Role =
```
[the remark this answers is not in the document]
Comment (tmlayton) [reply src/utilities/theme/types.ts]: Yes, good catch. This was originally `type Role = ...` and then

```ts
type RoleColor = {
  [r in Role]: string; 
};
```
but then decided to add the doc comments.

[the remark this answers is not in the document]
Comment (tmlayton) [reply src/utilities/theme/types.ts]: (or rather let’s pretend it was `Record<Role, string>` to satisfy your pettiness 😆)

File: src/utilities/theme/types.ts [hunk 4]
```diff
@@ -41,28 +63,20 @@ export interface Theme {
...
-  dark: HslaAdjustment;
+export interface Variant {
+  name: string;
+  description?: string;
+  light: HslaSetting;
+  dark: HslaSetting;
 }
 
-export type ColorAdjustments = Record<string, ColorRole>;
+export type RoleVariants = {
```
[the remark this answers is not in the document]
Comment (tmlayton) [reply src/utilities/theme/types.ts]: Well I like this pettiness

File: src/utilities/theme/utils.ts [hunk 5]
```diff
@@ -71,33 +86,49 @@ export function buildColors(theme: ThemeConfig) {
...
+              return {
+                ...acc3,
+                [displayName]: hslToString({
+                  ...colorToHsla(hsluvToHex([hue, saturation, lightness])),
+                  alpha,
+                }),
+              };
+            }, {}),
+          };
+        }, {}),
```
Comment (danrosenthal) [inline src/utilities/theme/utils.ts]: Some of the confusion around these nested reducers could be mitigated by using better names for the accumulators. It would also help understandability while refactoring.

Comment (tmlayton) [reply src/utilities/theme/utils.ts]: Yes, this will mostly go away when the factory is refactored to support child theme providers (only generating variants for a roles provided)

Comment (kyledurand) [reply src/utilities/theme/utils.ts]: Nice 👍

File: src/utilities/theme/types.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (BPScott) [reply src/utilities/theme/types.ts]: I guessed that might have been the case :) Loving the docblocks for the keys though, that's very tasty.
